cartoframes/magic.py

```diff
@@ -0,0 +1,82 @@
+"""carto magics"""
+from IPython.core.magic import (Magics, magics_class, line_magic,
+                                cell_magic, line_cell_magic)
+from cartoframes.context import CartoContext
+from cartoframes import utils
+# GOALS
+# - have it automatically find a CartoContext object
+# - make it agnostic to line or cell magics
+# - maps!
+
+def null_to_none(string):
+    """changes null to None"""
+    return None
+
+@magics_class
+class CartoMagics(Magics):
+    """Magics for bringing CARTO exploratory data analysis to IPython"""
+    @line_cell_magic
+    def carto(self, line, cell=None):
+        "query carto"
+        opts, table = self.parse_options(line, 'c:', posix=False, strict=False)
+        if cell:
+            query = cell
+        else:
+            query = 'select * from {}'.format(table)
+        context = getattr(opts, 'c', None)
+        if context is None:
+            # try to find a CartoContext instance if not specified
+            for key, val in self.shell.user_ns.items():
+                if isinstance(val, CartoContext):
+                    context = key
+                    break
+        if context is None:
+            raise ValueError('No CartoContext found or specified')
+        evalstr = "{0}.query(\'{1}\')".format(context, query.replace('\n', ' '))
+        return eval(evalstr, self.shell.user_ns)
+
+    @line_cell_magic
+    def cartomap(self, line, cell=None):
+        """carto map"""
+        opts, table = self.parse_options(line, 'c:s:ivt:',
+                                         posix=False, strict=False)
+        context = getattr(opts, 'c', None)
```

Comment:
Since `opts` seem to be `dict`s, we should use: `opts.get('c')` instead of getattr.

Reply:
I was just following documentation from http://ipython.readthedocs.io/en/stable/api/generated/IPython.core.magic.html#IPython.core.magic.Magics.parse_options, which says the `parse_options` is similar to `getopt.getopt()` except it returns a `Struct` with the options as keys instead of a list of key-value pairs. I wasn't sure if a `Struct` is the same as a `dict` but it certainly seems so!

Reply:
Perfect :) Glad you found something definitive on it.